src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -0,0 +1,1036 @@
+
+    // t3:a1 += (a2 * r0):t2
+    __ imulq(a2, r0);
+    __ addq(a1, t2);
+    __ adcq(t3, a2);
+  }
+
+  // At this point, 3 64-bit limbs are in t3:a1:a0
+  // t3 can span over more than 2 bits so final partial reduction step is needed.
+  //
+  // Partial reduction (just to fit into 130 bits)
+  //    a2 = t3 & 3
+  //    k = (t3 & ~3) + (t3 >> 2)
+  //         Y    x4  +  Y    x1
+  //    a2:a1:a0 += k
+  //
+  // Result will be in a2:a1:a0
+  __ movq(t1, t3);
+  __ movl(a2, t3); // DWORD
+  __ andq(t1, ~3);
+  __ shrq(t3, 2);
+  __ addq(t1, t3);
+  __ andl(a2, 3); // DWORD
+
+  // a2:a1:a0 += k (kept in t1)
+  __ addq(a0, t1);
+  __ adcq(a1, 0);
+  __ adcl(a2, 0); // DWORD
+}
+
+// Convert array of 128-bit numbers in quadwords (in D0:D1) into 128-bit numbers across 44-bit limbs (in L0:L1:L2)
+// Optionally pad all the numbers (i.e. add 2^128)
+//
+//         +-------------------------+-------------------------+
+//  D0:D1  | h0 h1 g0 g1 f0 f1 e0 e1 | d0 d1 c0 c1 b0 b1 a0 a1 |
+//         +-------------------------+-------------------------+
+//         +-------------------------+
+//  L2     | h2 d2 g2 c2 f2 b2 e2 a2 |
+//         +-------------------------+
+//         +-------------------------+
+//  L1     | h1 d1 g1 c1 f1 b1 e1 a1 |
+//         +-------------------------+
+//         +-------------------------+
+//  L0     | h0 d0 g0 c0 f0 b0 e0 a0 |
+//         +-------------------------+
+//
+void StubGenerator::poly1305_limbs_avx512(
+    const XMMRegister D0, const XMMRegister D1,
+    const XMMRegister L0, const XMMRegister L1, const XMMRegister L2, bool padMSG)
+{
+  const XMMRegister TMP1 = xmm0;
+  const XMMRegister TMP2 = xmm1;
+  const Register rscratch = r13;
+
+  // Interleave blocks of data
+  __ evpunpckhqdq(TMP1, D0, D1, Assembler::AVX_512bit);
+  __ evpunpcklqdq(L0, D0, D1, Assembler::AVX_512bit);
+
+  // Highest 42-bit limbs of new blocks
+  __ vpsrlq(L2, TMP1, 24, Assembler::AVX_512bit);
+  if (padMSG) {
+    __ evporq(L2, L2, ExternalAddress(poly1305_pad_msg()), Assembler::AVX_512bit, rscratch); // Add 2^128 to all 8 final qwords of the message
+  }
+
+  // Middle 44-bit limbs of new blocks
+  __ vpsrlq(L1, L0, 44, Assembler::AVX_512bit);
+  __ vpsllq(TMP2, TMP1, 20, Assembler::AVX_512bit);
+  __ vpternlogq(L1, 0xA8, TMP2, ExternalAddress(poly1305_mask44()), Assembler::AVX_512bit, rscratch); // (A OR B AND C)
+
+  // Lowest 44-bit limbs of new blocks
+  __ evpandq(L0, L0, ExternalAddress(poly1305_mask44()), Assembler::AVX_512bit, rscratch);
+}
+
+/**
+ * Copy 5×26-bit (unreduced) limbs stored at Register limbs into  a2:a1:a0 (3×64-bit limbs)
+ *
+ * a2 is optional. When only128 is set, limbs are expected to fit into 128-bits (i.e. a1:a0 such as clamped R)
+ */
+void StubGenerator::poly1305_limbs(const Register limbs, const Register a0, const Register a1, const Register a2, bool only128)
+{
+  const Register t1 = r13;
+  const Register t2 = r14;
+
+  __ movq(a0, Address(limbs, 0));
```

Comment:
I don't understand how it works. `limbs` comes directly from `c_rarg2` and contains raw oop. So, `Address(limbs, 0)` reads object mark word rather than the first element from the array. 

(Same situation in `poly1305_limbs_out`. And now I'm curious why doesn't object header corruption trigger a crash.)

Reply:
library_call.cpp takes care of that, it passes the address of 0'th element to the stub.

Reply:
Ah, got it. Worth elaborating that in the comments. Otherwise, they confuse rather than help:
```
  // void processBlocks(byte[] input, int len, int[5] a, int[5] r)
  const Register input        = rdi; //input+offset
  const Register length       = rbx;
  const Register accumulator  = rcx;
  const Register R            = r8;
```

Reply:
Added a comment, hopefully less confusing.

Reply:
On a second thought, passing derived pointers as arguments doesn't mix well with safepoint awareness.
(And this stub eventually has to become safepoint aware.)
Deriving a pointer inside the stub from a base oop and offset is trivial, recovering base oop from derived pointer is hard. 

It doesn't mean we have to address it right now.

Reply:
Left it as is. I also postponed Bytebuffer support for now, for a separate PR.. we can also fix it then?